File: 152-maximum-product-subarray/maximum-product-subarray.java
class Solution {
    public int maxProduct(int[] nums) {
       int res = nums[0];
       int curMax = nums[0];
       int curMin = nums[0];

       for( int n= 1; n<nums.length;n++){
            int i = nums[n];
            int temp = curMax *i;

            curMax = Math.max(Math.max(temp,curMin *i),i);
            curMin = Math.min(Math.min(temp,curMin *i),i);

            res = Math.max(res,curMax);            
       }
       return res;
    }
}

File: 152-maximum-product-subarray/maximum-product-subarray.py
class Solution:
    def maxProduct(self, nums: list[int]) -> int:
        if not nums:
            return 0
            
        res = max(nums)
        curMax = 1
        curMin = 1
        
        for n in nums:
            if n == 0:
                curMax = 1
                curMin = 1
                continue
                
            temp = curMax * n
            curMax = max(n * curMax, n * curMin, n)
            curMin = min(temp, n * curMin, n)
            res = max(res, curMax)
            
        return res
